unrouted_handler.go

```diff
@@ -437,12 +454,35 @@ func (handler *UnroutedHandler) PatchFile(w http.ResponseWriter, r *http.Request
 	}
 
 	// Do not proxy the call to the data store if the upload is already completed
-	if info.Offset == info.Size {
+	if !info.SizeIsDeferred && info.Offset == info.Size {
 		w.Header().Set("Upload-Offset", strconv.FormatInt(offset, 10))
 		handler.sendResp(w, r, http.StatusNoContent)
 		return
 	}
 
+	if r.Header.Get("Upload-Length") != "" {
+		if handler.composer.UsesLengthDeferrer {
+			if info.SizeIsDeferred {
+				uploadLength, err := strconv.ParseInt(r.Header.Get("Upload-Length"), 10, 64)
+				if err != nil || uploadLength < 0 || uploadLength < info.Offset || uploadLength > handler.config.MaxSize {
+					handler.sendError(w, r, ErrInvalidUploadLength)
+					return
+				}
+
+				info.Size = uploadLength
+				info.SizeIsDeferred = false
+				if err := handler.composer.LengthDeferrer.DeclareLength(id, info.Size); err != nil {
+					handler.sendError(w, r, err)
+					return
+				}
+			} else {
+				handler.sendError(w, r, ErrInvalidUploadLength)
+			}
+		} else {
+			handler.sendError(w, r, ErrNotImplemented)
```

Comment:
I think christmas trees like these should be avoided if reasonably possible by returning home early. I think there's an official term for it, but I couldn't find it. Maybe this post that I did find helps illustrate the concept tho http://blog.timoxley.com/post/47041269194/avoid-else-return-early :)

Reply:
i realize this pr has been merged already and it's nothing critical. just a friendly tip for future endeavours :)

Reply:
Thanks for the feedback, @kvz. I agree -- an early return would be better. I also realized that this PR doesn't add `Upload-Defer-Length` to the CORS allowed header list, so I'll create a followup PR that addresses both.